### ISSUE TITLE
Pass arguments from the cli

### ISSUE THREAD

**Issue by thufschmitt** [tweag/nickel#413]
**Is your feature request related to a problem? Please describe.**

It would be very nice to allow nickel scripts to be easily parametrized. For example one could imagine a nickel script deploying a k8s cluster, and have the user want to easily specify how many instances to start.

**Describe the solution you'd like**

Allow passing arguments to nickel scripts from the CLI. A simple way could be to have a `--arg` flag such that `nickel -f foo.ncl --arg 'someNickelExpr'` would be equivalent to `nickel <<< '(import "foo.ncl") someNickelExpr'` (or if nickel were to have named arguments, have a nix-style `--arg name value` flag)

**Describe alternatives you've considered**

- Explicitely use `nickel <<< '(import "foo.ncl") blah'` (possibly wrapping that behind a script that would do the interpolation). That works but that’s a bit painful and fragile
- Have a nickel builtin able to query a value from the environment (from an environment variable probably). That would work rather nicely, but it’s also a great source of hard-to-track non-determinisim.

**Comment by marziply** [comment]
I've just stumbled upon Nickel in an effort to find a configuration language I can use to replace Kustomize which has several limitations that make it incompatible with the infrastructure I'm currently working on. The biggest issue currently is that Kustomize does not accept any parameter inputs or environment variables, limiting the options I have for multi environment deployments. I can separate my deployments into stages as directories (production, development, etc.), but this doesn't work well if I have multiple kinds of deployments each with multiple environmental stages. Plus which, I'm unable to configure particular properties within these deployments that might differ slightly between each other. Not without something like `envsubst` which isn't particularly ergonomic for devs that just want to consistently execute `kubectl ... -k ...` without having to substitute environment variables.

What this is leading me to say is this. Without input parameters or support for environment variables, this project limits itself on how exactly configuration is applied. Sure, there are functions, validation, and lots of DRY. Without input parameters though, those features don't hold much weight when I've got YAML anchors, map generators, transformers, and the many other conveniences Kustomize provides.

This configuration language looks great, and it has a rich feature set that looks appealing on the surface. For me personally though, those features simply aren't worth anything without input parameters. I don't want to execute Nickel with a `import "..."` argument to determine how my configuration is generated. That feels way too analogous to Kustomize staging directories which has several limitations, one of which being what I've described above.

In summary, I'm keen on using Nickel, but until there is support for input parameters or environment variables, I'm going to be sticking with Kustomize.

**Comment by marziply** [comment]
My use case is a replacement for Kustomize. The limitation I've glossed over previously can be better described like so:

Let's assume I have a monorepo of various customer deployments that all share the same service configurations except with subtle variations dependant on the customers wishes.

```
src
`- customer_a
   `- production
      `- deployment.yaml
      `- kustomization.yaml
   `- staging
      `- deployment.yaml
      `- kustomization.yaml
   `- development
      `- deployment.yaml
      `- kustomization.yaml
`- customer_b
   `- production
      `- deployment.yaml
      `- kustomization.yaml
   `- staging
      `- deployment.yaml
      `- kustomization.yaml
   `- development
      `- deployment.yaml
       `- kustomization.yaml
`- customer_c
   `- production
      `- deployment.yaml
      `- kustomization.yaml
   `- staging
      `- deployment.yaml
      `- kustomization.yaml
   `- development
      `- deployment.yaml
      `- kustomization.yaml
```

Given this contrived example, there isn't an ergonomic solution with Kustomize to permit simplifying the above into:

```
src
`- customers
   `- production
      `- deployment.yaml
      `- kustomization.yaml
   `- staging
      `- deployment.yaml
      `- kustomization.yaml
   `- development
      `- deployment.yaml
      `- kustomization.yaml
```

Assuming both examples above all share identical `deployment.yaml` except for maybe `image: foo/bar:production`, `image: foo/bar:development`, `args: ["--customer", <customer-name>]`, etc., I would prefer to have the option of defining those differences as input parameters rather than hard coded dependant on the file system.

Instead of using Kustomize, I could opt for tooling such as Terraform or Helm. They both support input parameters and easily solve the issue above. Although they both have issues of their own, each of which I have explored in other unrelated projects. The reason I'm more keen on Nickel instead is primarily for the Turing completeness that's lacking in configurations languages such as HCL.

I think both of your suggestions are worth exploring. Personally I would prefer `nickel -f "config.ncl" --arg customer acme` as that seems to be the most ergonomic in my opinion. Although I would see no harm in perhaps having both?

**Comment by yannham** [comment]
Superseded by #1408 and fixed by #1475 (CLI customize mode, shipped in latest 1.2.1)
